staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go

```diff
@@ -161,3 +190,206 @@ func RecoverFromPanic(err *error) {
+// If the exact key does not exist in d.count, a similar enough fuzzy match on key.message
+// will be used as a fallback (the stack and error type must always match via ==).
+// If fuzzy matching is used, the given key's message will be updated to the similar message.
+// Thus this method requires that key be passed in as a pointer.
+func (d *dedupingErrorHandler) getVal(key *errKey) (errVal, bool) {
+	val, isOldErr := d.count[*key]
+	// found direct match, use that before doing levenshtein fuzzy lookup
+	if isOldErr {
+		return val, false // return false because this is not a new error
+	}
+
+	// Do not bother doing fuzzy checks if the direct match failed and we are set to require identical strings
+	// TODO remove this once we have a useful default value for similar
+	if d.Similar >= 1 {
+		return errVal{}, true // return true because this is a new error
+	}
+
+	var (
+		// fuzzySimilarity is initialized to d.similar since that is the
+		// lowest similarity value we consider as the strings being "equal"
+		fuzzySimilarity = d.Similar
+		fuzzyFound      bool
+		fuzzyMessage    string
+		fuzzyVal        errVal
+	)
+
+	// we have to iterate over the whole map to do fuzzy matching on errKey.message
+	// this is ok because d.count should always be relatively small
+	// note that we cannot exit this loop early since we want to pick the most similar message
+	for k, v := range d.count {
+		// the stack and error type must match
+		if key.stack == k.stack && key.errType == k.errType {
+			// now we perform fuzzy matching on the message
+			if s := levenshtein.Similarity(key.message, k.message, nil); s > fuzzySimilarity {
+				fuzzySimilarity = s
+				fuzzyFound = true
+				fuzzyMessage = k.message
+				fuzzyVal = v
+			}
+		}
+	}
+
+	if fuzzyFound {
+		// update the input key's message to match our fuzzy search
+		key.message = fuzzyMessage
+		return fuzzyVal, false // return false because we do not consider this a new error
+	}
+
+	return errVal{}, true // return true because this is a new error
+}
+
+// logError uses glog to log at the call site of HandleError
+// it must be called from dedupingErrorHandler.handleErr
+func (d *dedupingErrorHandler) logError(err error, count uint64) {
+	// get the full name of the file since glog likes to strip it
+	// TODO if we are ok with multi line messages, we could simply print the whole stack
+	_, file, line, ok := runtime.Caller(d.errorDepth)
+	if !ok {
+		file = "unknown_file"
+		line = -1
+	}
+	glog.ErrorDepth(d.errorDepth, fmt.Sprintf("%s:%d err=%v count=%d value=%#v", file, line, err, count, err))
+}
+
+var (
+	hexNumberRE  = regexp.MustCompile(`0x[0-9a-f]+`)
+	emptyAddress = []byte("0x?")
+)
+
+// getStack returns the important part of the stack trace
+// it must be called from dedupingErrorHandler.handleErr
+// it must not mutate d (or even read d.cache/d.count) since it is called when the lock is not held
+func (d *dedupingErrorHandler) getStack() string {
```

Comment:
I'd like to see 2 tests, one which has golden data and shows that this works, and one that captures a live stack trace and will fail if the output of that function ever changes.